File: src/grupo_1/equipo_5_seminario_14/Code/Decorator Pattern implementations/decorator_pattern.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Decorator
{
    class Program
    {
        static void Main(string[] args)
        {
            BotonDeAyudaDecorator ventanaConBotonDeAyuda = new BotonDeAyudaDecorator(new Ventana());
            BordeDecorator ventanaConBotonDeAyudaYBorde = new BordeDecorator(ventanaConBotonDeAyuda);
            ventanaConBotonDeAyudaYBorde.Dibujar();
            Console.WriteLine();

            BordeDecorator ventanaConBorde = new BordeDecorator(new Ventana());
            ventanaConBorde.Dibujar();
            Console.WriteLine();

            BordeDecorator ventanaConDobleBorde = new BordeDecorator(ventanaConBorde);
            ventanaConDobleBorde.Dibujar();
            Console.WriteLine();
            
            BordeDecorator ventanaConDobleBordeYBotonDeAyuda = new BordeDecorator(new BordeDecorator(ventanaConBotonDeAyuda));
            ventanaConDobleBordeYBotonDeAyuda.Dibujar();
            Console.WriteLine();
            Console.Read();
        }
        //[Clase Component] ver diagrama de clases
        public interface IVentanaAbstracta
        {
            void Dibujar();
        }
        //[Clase ConcreteComponent] ver diagrama de clases, Clase que se desea decorar
        public class Ventana : IVentanaAbstracta
        {
            public void Dibujar() { Console.Write(" Ventana "); }
        }
        //[Clase Decorator] ver diagrama de clases
        public abstract class VentanaDecorator : IVentanaAbstracta
        {
            public VentanaDecorator(IVentanaAbstracta ventanaAbstracta) { _VentanaAbstracta = ventanaAbstracta; }
            protected IVentanaAbstracta _VentanaAbstracta;
            public abstract void Dibujar();
        }
        //[Clase ConcreteDecorator] ver diagrama de clases
        public class BordeDecorator : VentanaDecorator
        {
            public BordeDecorator(IVentanaAbstracta ventanaAbstracta) : base(ventanaAbstracta) { }
            public override void Dibujar() { Console.Write("|"); _VentanaAbstracta.Dibujar(); Console.Write("|"); }            
        }
        //[Clase ConcreteDecorator] ver diagrama de clases
        public class BotonDeAyudaDecorator : VentanaDecorator
        {
            public BotonDeAyudaDecorator(IVentanaAbstracta ventanaAbstracta) : base(ventanaAbstracta) { }
            public override void Dibujar() { _VentanaAbstracta.Dibujar(); Console.Write("[Boton de Ayuda]"); }            
        }
    }
}

File: src/grupo_1/equipo_5_seminario_14/Codigo/Decorator Pattern implementations/decorator_pattern.cpp
#include <iostream>

//[Clase Component] ver diagrama de clases
class IVentanaAbstracta
{
public:
    virtual void Dibujar() = 0;
};

//[Clase ConcreteComponent] ver diagrama de clases, Clase que se desea decorar
class Ventana : public IVentanaAbstracta
{
public:
    void Dibujar()
    {
        std::cout << " Ventana ";
    }
};

//[Clase Decorator] ver diagrama de clases
class VentanaDecorator : public IVentanaAbstracta
{
public:
    VentanaDecorator(IVentanaAbstracta *ventanaAbstracta)
    {
        _VentanaAbstracta = ventanaAbstracta;
    }

    virtual void Dibujar() = 0;

protected:
    IVentanaAbstracta *_VentanaAbstracta;
};

//[Clase ConcreteDecorator] ver diagrama de clases
class BordeDecorator : public VentanaDecorator
{
public:
    BordeDecorator(IVentanaAbstracta *ventanaAbstracta) : VentanaDecorator(ventanaAbstracta)
    {
    }

    virtual void Dibujar()
    {
        std::cout << "|";
        _VentanaAbstracta->Dibujar();
        std::cout << "|";
    }
};

//[Clase ConcreteDecorator] ver diagrama de clases
class BotonDeAyudaDecorator : public VentanaDecorator
{
public:
    BotonDeAyudaDecorator(IVentanaAbstracta *ventanaAbstracta) : VentanaDecorator(ventanaAbstracta)
    {
    }

    virtual void Dibujar()
    {
        _VentanaAbstracta->Dibujar();
        std::cout << "[Boton de Ayuda]";
    }
};

int main()
{
    BotonDeAyudaDecorator *ventanaConBotonDeAyuda = new BotonDeAyudaDecorator(new Ventana());
    ventanaConBotonDeAyuda->Dibujar();
    std::cout << std::endl;

    BordeDecorator *ventanaConBotonDeAyudaYBorde = new BordeDecorator(ventanaConBotonDeAyuda);
    ventanaConBotonDeAyudaYBorde->Dibujar();
    std::cout << std::endl;

    BordeDecorator *ventanaConBorde = new BordeDecorator(new Ventana());
    ventanaConBorde->Dibujar();
    std::cout << std::endl;

    BordeDecorator *ventanaConDobleBorde = new BordeDecorator(ventanaConBorde);
    ventanaConDobleBorde->Dibujar();
    std::cout << std::endl;

    delete ventanaConBotonDeAyuda;
    delete ventanaConBotonDeAyudaYBorde;
    delete ventanaConBorde;
    delete ventanaConDobleBorde;

    return 0;
}